src/worker/jobs/downloads/update_metadata.rs

```diff
@@ -17,24 +18,52 @@ impl BackgroundJob for UpdateDownloads {
     async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
         spawn_blocking(move || {
             let mut conn = env.connection_pool.get()?;
-            update(&mut conn)?;
+            update(&mut conn, env.config.batch_update_downloads)?;
             Ok(())
         })
         .await
     }
 }
 
-fn update(conn: &mut PgConnection) -> QueryResult<()> {
+fn update(conn: &mut PgConnection, use_batch_update: bool) -> QueryResult<()> {
     use diesel::dsl::now;
     use diesel::select;
 
-    let rows = version_downloads::table
-        .filter(version_downloads::processed.eq(false))
-        .filter(version_downloads::downloads.ne(version_downloads::counted))
-        .load(conn)?;
-
-    info!(rows = rows.len(), "Updating versions");
-    collect(conn, &rows)?;
+    if use_batch_update {
+        info!("Updating versions…");
+
+        // After 45 minutes, we stop the batch updating process to avoid
+        // triggering the long-running job alert. 15 minutes later, the job
+        // will be started again by our cron service anyway.
+        const TIME_LIMIT: Duration = Duration::from_secs(45 * 60);
```

Comment:
Do we actually want to do this? If this job is unable to keep up with the uncounted rows, is there an alert that would fire?

Reply:
The way this worked previously is that on job startup it queried the full list of rows that need updates and then worked through them until all have been processed.

With this new query we are constantly querying new data too, which could lead to a situation where the loop never quits. From my estimations we probably should be done with this job in 3-5 min. with the new implementation, but just to be save I've added this limit.

I'd feel a bit uneasy if we used an infinite loop that could actually run forever, but I'm open to other suggestions :)

Reply:
That's a good point, thanks. I think I still had the old behaviour of calculating the update set in advance in my head when I wrote that.

That said, I still think we'll have to figure out some sort of alerting strategy here in the longer term to avoid the problem where the job just falls further and further behind. I'm not sure exactly what that looks like before I'd had coffee at 7:39 am.

Reply:
just as a data point: the runs so far have taken roughly 5-10 minutes, though sometimes interrupted by the deadlock issues I had mentioned

Reply:
That's good, then. Eventually this will probably be an issue, but presumably something else will break first.